internal/worker_versioning_rules.go

```diff
@@ -0,0 +1,338 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"errors"
+	"time"
+
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	// VersioningRamp is an interface for the different strategies of gradual workflow deployments.
+	VersioningRamp interface {
+		isValidVersioningRamp() bool
+	}
+
+	// VersioningRampByPercentage sends a proportion of the traffic to the target Build ID.
+	VersioningRampByPercentage struct {
+		// Percentage of traffic with a value in [0,100)
+		Percentage float32
+	}
+
+	// VersioningAssignmentRule is a BuildID assigment rule for a task queue.
+	// Assignment rules only affect new workflows.
+	VersioningAssignmentRule struct {
+		// The BuildID of new workflows affected by this rule.
+		TargetBuildId string
+		// A strategy for gradual workflow deployment.
+		Ramp VersioningRamp
+		// An optional registration time set by the server.
+		Timestamp *time.Time
+	}
+
+	// VersioningAssignmentRule is a BuildID redirect rule for a task queue.
+	// It changes the behavior of currently running workflows and new ones.
+	VersioningRedirectRule struct {
+		SourceBuildId string
+		TargetBuildId string
+		// An optional registration time set by the server.
+		Timestamp *time.Time
+	}
+
+	//VersioningConflictToken is a conflict token to serialize updates.
+	// An update with an old token fails with `serviceerror.FailedPrecondition`.
+	// The current token can be obtained with [GetWorkerVersioningRules], or returned by a successful [UpdateWorkerVersioningRules].
+	VersioningConflictToken struct {
+		token []byte
+	}
+
+	// UpdateWorkerVersioningRulesOptions is the input to Client.UpdateWorkerVersioningRules.
+	UpdateWorkerVersioningRulesOptions struct {
+		// The task queue to update the versioning rules of.
+		TaskQueue string
+		// A conflict token to serialize updates.
+		ConflictToken VersioningConflictToken
+		Operation     UpdateVersioningOp
+	}
+
+	// UpdateVersioningOp is an interface for the different operations that can be
+	// performed when updating the worker versioning rules for a task queue.
+	//
+	// Possible operations are:
+	//   - VersioningOpInsertAssignmentRule
+	//   - VersioningOpReplaceAssignmentRule
+	//   - VersioningOpDeleteAssignmentRule
+	//   - VersioningOpInsertRedirectRule
+	//   - VersioningOpReplaceRedirectRule
+	//   - VersioningOpDeleteRedirectRule
+	//   - VersioningOpCommitBuildId
+	UpdateVersioningOp interface {
```

Comment:
Seems like this should just be `VersioningOp`, though consider calling it `VersioningOperation` (no reason to shorten the last word I don't think)

Reply:
Yes, let's be consistent with the others.